apps/guardian/src-tauri/src/main.rs

```diff
@@ -1029,7 +1046,363 @@ async fn generate_secret_share_flow<'a>(
+    // returned if either:
+    //
+    // - Shares received from peers cannot be combined
+    // - Decryption proof cannot be generated for some ciphertext
+    let generate_decryption_proof = |proof_data: &ProofData| -> Result<_, Error> {
+        let (commit_shares, decryption_shares): (Vec<_>, Vec<_>) = proof_data
+            .decryption_shares
+            .iter()
+            .map(|share| (share.proof_commit.clone(), &share.share))
+            .unzip();
+
+        let combined_decryption = CombinedDecryptionShare::combine(&parameters, decryption_shares)
+            .map_err(|e| {
+                eprintln!("{e}");
+                Error::DecryptionShareError("Failed to combine shares received from peers".into())
+            })?;
+        let proof = DecryptionProof::generate_response_share(
+            &parameters.fixed_parameters,
+            &context.hashes_ext,
+            &context.public_key,
+            proof_data.ciphertext,
+            &combined_decryption,
+            &commit_shares,
+            proof_data.secret_state,
+            &secret_key_share,
+        )
+        .map_err(|e| {
+            eprintln!("{e}");
+            Error::DecryptionShareError("Failed to generate the response share to register".into())
+        })?;
+
+        Ok(proof)
+    };
+
+    // A map of ciphertexts paired with associated secret state + guardian shares
+    // for each contest in the election. Errors are returned if:
+    //
+    // - Decryption shares do not match the format of the tally
+    let proof_data: BTreeMap<_, _> = encrypted_tally
+        .iter()
+        .map(|(contest_index, ciphertexts)| {
+            let secret_states = match secret_states.get(contest_index) {
+                Some(secret_states) if secret_states.len() == ciphertexts.len() => secret_states,
+                _ => return Err(anyhow!("Invalid secret states for tally")),
+            };
+            let decryption_shares: Vec<_> = decryption_shares
+                .iter()
+                .map(|shares| match shares.get(contest_index) {
+                    Some(shares) if shares.len() == ciphertexts.len() => Ok(shares),
+                    _ => Err(Error::DecryptionShareError(
+                        "Invalid decryption shares detected".into(),
+                    )),
+                })
+                .try_collect()?;
+
+            // For ciphertexts in a contest, find the corresponding decryption share from
+            // each guardian
+            let proof_data = ciphertexts
+                .iter()
+                .zip(secret_states)
+                .enumerate()
+                .map(|(i, (ciphertext, secret_state))| {
+                    let decryption_shares =
+                        decryption_shares
+                            .iter()
+                            .fold(Vec::new(), |mut acc, shares| {
+                                acc.push(&shares[i]);
+                                acc
+                            });
```

Comment:
```suggestion
                            .iter()
                            .map(|shares| {
                                &shares[i])
                            }).collect();
```

Also, why do we know `shares[i]` will not panic?

Reply:
Prior to this (line 1288 and 1294), a check is made to ensure the length of the entries in all three parts needed for the proof (i.e. the ciphertexts, the "secret states" and the decryption shares registered in the contract). I'll write a comment